### ISSUE TITLE
Unexpected Docker Daemon shutdown due to ServeAPI error "The pipe is being closed" when docker stats was used

### ISSUE THREAD

**Issue by georgyturevich** [moby/moby#34207]
**Description**

Docker Daemon stopped when I tried to use following command:
```
docker stats --no-stream --format 'table{{.BlockIO}}\t{{.Name}}' $(docker ps -qa)
```

**Steps to reproduce the issue:**
1. Restart Docker daemon in debug mode
```
Stop-Service docker
dockerd -D > daemon.log 2>&1
```

2. Pull microsoft/iis image if you do not have it
```
docker pull microsoft/iis;
```

3. In first Powershell console fun script which starts 100 iis containers and then simulate some activity by restarting containers in an infinite loop
```
for ($i=0; $i -le 100; $i++) {`
	$cname = "test_iis$i"; `
	docker stop $cname; `
	docker rm $cname; `
	$ExtPort = 10000 + $i;`
	docker run -d --name $cname -p $ExtPort`:80 microsoft/iis ping -t localhost; `
}`
$K=0;`
while($True) {` 
	$K++;
	$M=0;`
	$Containers = $(docker ps --format "{{.Names}}");`
	$Count = ($Containers).count;`
	$Containers | foreach { `
		$M++; `
		$Start = Get-Date; echo "$Start`: Restarting #$M/$K (of $Count) $_ ..."; `
		docker restart $_; `
		$Stop = Get-Date; echo "$stop`: #$M (of $Count) $_ started in $(($Stop - $Start).TotalSeconds) s."; `
	}`
}
```

4. In second Powershell console run another script which executes `docker stats` command in an infinite loop
```
$i=0;`
while($True) {` 
	$i++; echo "$(date): #$i"; docker stats --no-stream --format 'table{{.BlockIO}}\t{{.Name}}' $(docker ps -qa);` 
}
```
5. After all 100 containers are started during next 10-20 minutes Docker Daemon will be stopped. 

Inside daemon.log I see following information:

```
time="2017-07-20T19:25:47.500023900Z" level=debug msg="Calling GET /v1.27/containers/cc9d64ef144e/stats?stream=0" 

... (hundreds of stats requests)
time="2017-07-20T19:25:47.572026200Z" level=debug msg="Calling GET /v1.27/containers/2c8d33c286b1/stats?stream=0" 
time="2017-07-20T19:25:47.573027100Z" level=debug msg="Calling GET /v1.27/containers/e558e6949cec/stats?stream=0" 
time="2017-07-20T19:25:47.574026600Z" level=error msg="ServeAPI error: The pipe is being closed." 
time="2017-07-20T19:25:47.578027500Z" level=debug msg="start clean shutdown of all containers with a 15 seconds timeout..." 
...
time="2017-07-20T20:52:36.830675700Z" level=error msg="Force shutdown daemon" 
dockerd : Shutting down due to ServeAPI error: The pipe is being closed.
```

Interesting thing that if I run container without exposing ports (`-p $ExtPort:80`) then I can not reproduce this issue quickly. So may be it is related somehow to HNS/networking.

Initially this issue was caught on Docker Daemon `17.03.1-ee-3` and then reproduced by me with `17.06.0-ce`

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `docker version`:**

```
Client:
 Version:      17.06.0-ce
 API version:  1.30
 Go version:   go1.8.3
 Git commit:   02c1d87
 Built:        Fri Jun 23 21:30:30 2017
 OS/Arch:      windows/amd64

Server:
 Version:      17.06.0-ce
 API version:  1.30 (minimum version 1.24)
 Go version:   go1.8.3
 Git commit:   02c1d87
 Built:        Fri Jun 23 22:19:00 2017
 OS/Arch:      windows/amd64
 Experimental: false
```

**Output of `docker info`:**

```
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 14
Server Version: 17.06.0-ce
Storage Driver: windowsfilter
 Windows:
Logging Driver: json-file
Plugins:
 Volume: local
 Network: l2bridge l2tunnel nat null overlay transparent
 Log: awslogs etwlogs fluentd json-file logentries splunk syslog
Swarm: inactive
Default Isolation: process
Kernel Version: 10.0 14393 (14393.1358.amd64fre.rs1_release.170602-2252)
Operating System: Windows Server 2016 Datacenter
OSType: windows
Architecture: x86_64
CPUs: 4
Total Memory: 16GiB
Name: EC2AMAZ-N2GM0M9
ID: LH3O:5PAX:ATLN:NPF3:THQV:MUQ4:ZIPT:SAYQ:RV53:SG3B:QXGT:6C2I
Docker Root Dir: E:\docker_storage_1_13
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: -1
 Goroutines: 20
 System Time: 2017-07-20T20:39:05.3530276Z
 EventsListeners: 0
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
It is AWS m4.xlarge instance (1 socket, 4 virtual CPUs, 16 Gib memory)

**Comment by friism** [comment]
cc @jhowardmsft @johnstep

**Comment by georgyturevich** [comment]
@friism Hi Michael,

I was able to reproduce this issue with `17.09-ce` and the same test case. Attaching full log file
[daemon.log.zip](https://github.com/moby/moby/files/1380501/daemon.log.zip)


**Output of `docker version`:**

```
Client:
 Version:      17.09.0-ce
 API version:  1.32
 Go version:   go1.8.3
 Git commit:   afdb6d4
 Built:        Tue Sep 26 22:40:09 2017
 OS/Arch:      windows/amd64

Server:
 Version:      17.09.0-ce
 API version:  1.32 (minimum version 1.24)
 Go version:   go1.8.3
 Git commit:   afdb6d4
 Built:        Tue Sep 26 22:50:27 2017
 OS/Arch:      windows/amd64
 Experimental: false
```

**Output of `docker info`:**

```
Containers: 101
 Running: 0
 Paused: 0
 Stopped: 101
Images: 1
Server Version: 17.09.0-ce
Storage Driver: windowsfilter
 Windows:
Logging Driver: json-file
Plugins:
 Volume: local
 Network: ics l2bridge l2tunnel nat null overlay transparent
 Log: awslogs etwlogs fluentd json-file logentries splunk syslog
Swarm: inactive
Default Isolation: process
Kernel Version: 10.0 14393 (14393.1715.amd64fre.rs1_release_inmarket.170906-1810)
Operating System: Windows Server 2016 Datacenter
OSType: windows
Architecture: x86_64
CPUs: 4
Total Memory: 16GiB
Name: EC2AMAZ-V4RG8E1
ID: VZIV:QABT:63VB:KUOW:4FOR:BTTB:QULZ:U7KS:CIZK:QWTU:63SI:65EG
Docker Root Dir: C:\ProgramData\docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Comment by georgyturevich** [comment]
I was able to reproduce similar issue but with `docker stop` command only https://github.com/moby/moby/issues/35183. May be these issues has a common nature.

**Comment by jhowardmsft** [comment]
OK, I can repro on a big Azure machine. I'll see what I can find.

**Comment by ilyalukyanov** [comment]
Hey @jhowardmsft have you managed to discover anything about the bug?

We're running a `Kubernetes 1.7.5` cluster with 4 `Windows Server 2016 Datacenter` nodes running `docker 17.06.1-ee-2` hosted in Azure. Cluster is provisioned using `acs-engine` built from sources. `acs-engine` returns:
```
Version: canary
GitCommit: 8d488170
GitTreeState: clean
```

For some reason even though all 4 nodes are identical, the issue seems to be localised to one node only or at least it happens there way more frequently. The cluster had been running for a month or so until about a week ago. 

A few days before the first crash we deployed additional ~20 containers to the cluster. However, the issue doesn't seem to be related to load as once it happened when the node wasn't running any containers. That could be caused by some calls K8S makes to docker.


We're suffering very badly from it and I'd be keen to help if there's anything I could do.

**Comment by jhowardmsft** [comment]
See #35968

**Comment by jbiel** [comment]
This is happening to us on a test Kubernete Windows cluster with 3 nodes. Usage is very light at this point (only one container/pod per node) and crash frequency is high (we lose approx one node per day.)

If #35968 fixes this it'd be great to see that patch quickly make its way into a Docker EE build.

Environment:
* Windows Server 2016 R3 (1709)
* Kubernetes 1.9.4
```Server:
$ docker version
 Version:      17.06.2-ee-6
 API version:  1.30 (minimum version 1.24)
 Go version:   go1.8.3
 Git commit:   e75fdb8
 Built:        Mon Nov 27 22:55:16 2017
 OS/Arch:      windows/amd64
 Experimental: true

**Comment by jhowardmsft** [comment]
@carlfisher1 ^^

**Comment by jhowardmsft** [comment]
@carlfischer1 I meant. &&

**Comment by carlfischer1** [comment]
@jbiel I've verified we have a tracking issue for potential EE backport of #35968

**Comment by jbiel** [comment]
@carlfischer1 thanks! I assume that the repo/project for Docker EE is private?

**Comment by carlfischer1** [comment]
@jbiel it is. will post back to this thread with updates.

**Comment by georgyturevich** [comment]
@carlfischer1 Hi Carl,

Do you have any news about a backport of this issue?

Thanks!

**Comment by jorisscheppers** [comment]
Very interested in which EE version this fix ends up in. Our development Docker host has the problem of unexpected daemon shutdowns too.

**Comment by georgyturevich** [comment]
@jorisscheppers The fix was included to 17.06.2-ee-14 (2018-06-21) according to the release notes https://docs.docker.com/ee/engine/release-notes/#runtime-3

I think 18.03.1-ee also contains the fix.
